src/cached_object_store.rs

```diff
@@ -762,6 +817,260 @@ impl LocalCacheEntry for FsCacheEntry {
+        let iter = WalkDir::new(&self.root_folder).into_iter();
+        for entry in iter {
+            let entry = match entry {
+                Ok(entry) => entry,
+                Err(err) => {
+                    warn!("evictor: failed to walk the cache folder: {}", err);
+                    continue;
+                }
+            };
+            if entry.file_type().is_dir() {
+                continue;
+            }
+
+            let metadata = match tokio::fs::metadata(entry.path()).await {
+                Ok(metadata) => metadata,
+                Err(err) => {
+                    warn!(
+                        "evictor: failed to get the metadata of the cache file: {}",
+                        err
+                    );
+                    continue;
+                }
+            };
+            let atime = metadata.accessed().unwrap_or(SystemTime::UNIX_EPOCH);
+            let path = entry.path().to_path_buf();
+            let bytes = metadata.len();
+
+            self.cache_entries
+                .lock()
+                .await
+                .insert(path, (atime, bytes as usize));
+            self.cache_size_bytes.fetch_add(bytes, Ordering::SeqCst);
+        }
+    }
+
+    pub async fn maybe_evict(&self, path: std::path::PathBuf, bytes: usize) -> usize {
+        // record the new cache entry into the cache_entries, and increase the cache_size_bytes
+        self.cache_size_bytes
+            .fetch_add(bytes as u64, Ordering::SeqCst);
+        self.cache_entries
+            .lock()
+            .await
+            .insert(path.clone(), (SystemTime::now(), bytes));
+
+        // record the metrics
+        self.db_stats
+            .object_store_cache_keys
+            .set(self.cache_entries.lock().await.len() as u64);
+        self.db_stats
+            .object_store_cache_bytes
+            .set(self.cache_size_bytes.load(Ordering::Relaxed));
+
+        // if the cache size is still below the limit, do nothing
+        if self.cache_size_bytes.load(Ordering::Relaxed) <= self.max_cache_size_bytes as u64 {
+            return 0;
+        }
+        // TODO: check the disk space ratio here, if the disk space is low, also triggers evict.
+
+        // if the cache size exceeds the limit, evict the cache files in batch with the batch_factor,
+        // this may help to avoid the cases like triggering the evictor too frequently when the cache
+        // size is just slightly above the limit.
+        let mut total_bytes: usize = 0;
+        for _ in 0..self.batch_factor {
+            let evicted_bytes = self.maybe_evict_once().await;
+            if evicted_bytes == 0 {
+                return total_bytes;
+            }
+
+            self.db_stats
+                .object_store_cache_evicted_bytes
+                .add(evicted_bytes as u64);
+            self.db_stats.object_store_cache_evicted_keys.inc();
+
```

Comment:
Should `object_store_cache_keys` and `object_store_cache_bytes` be set/decremented here? Otherwise, it seems the metrics are inaccurate until the next `maybe_evict` is called.

Reply:
makes sense, made the change in 6e10adc